File: apis/pkg/types/environment_promotion.go
package types

import "time"

// EnvironmentPromotionTopology is the editable global environment promotion graph.
// Only baseline environments participate in Edges. Lanes return changes through
// LaneBaseBindings so that the baseline graph remains acyclic.
type EnvironmentPromotionTopology struct {
	DraftRevision     uint64                     `json:"draft_revision"`
	PublishedRevision uint64                     `json:"published_revision"`
	Edges             []EnvironmentPromotionEdge `json:"edges"`
	LaneBaseBindings  []LaneBaseBinding          `json:"lane_base_bindings"`
	ModifyBy          string                     `json:"modify_by,omitempty"`
	ModifyTime        time.Time                  `json:"modify_time,omitempty"`
}

// EnvironmentPromotionEdge defines one allowed baseline promotion direction.
type EnvironmentPromotionEdge struct {
	ID              string   `json:"id"`
	Source          string   `json:"source"`
	Target          string   `json:"target"`
	RequireFormal   bool     `json:"require_formal_release"`
	RequireApproval bool     `json:"require_approval"`
	ValidationGates []string `json:"validation_gates,omitempty"`
	AllowedDomains  []string `json:"allowed_resource_domains,omitempty"`
	ConflictPolicy  string   `json:"conflict_policy,omitempty"`
}

// LaneBaseBinding binds a lane to exactly one baseline environment.
type LaneBaseBinding struct {
	Lane string `json:"lane"`
	Base string `json:"base"`
}

// EnvironmentPromotionTopologyRevision is an immutable published topology snapshot.
type EnvironmentPromotionTopologyRevision struct {
	Revision   uint64                       `json:"revision"`
	Topology   EnvironmentPromotionTopology `json:"topology"`
	Comment    string                       `json:"comment,omitempty"`
	CreateBy   string                       `json:"create_by,omitempty"`
	CreateTime time.Time                    `json:"create_time,omitempty"`
}

// TopologyValidationIssue is stable enough for both API and Console field feedback.
type TopologyValidationIssue struct {
	Code    string `json:"code"`
	Path    string `json:"path,omitempty"`
	Message string `json:"message"`
}

File: web/console/src/services/environment_promotion.ts
import { apiRequest, getApiRequest, putApiRequest } from 'utils/request';

const TOPOLOGY_URL = '/core/v1/environment-promotion/topology';

export interface EnvironmentPromotionEdge {
  id: string;
  source: string;
  target: string;
  require_formal_release: boolean;
  require_approval: boolean;
  validation_gates: string[];
  allowed_resource_domains: string[];
  conflict_policy: 'BLOCK';
}

export interface LaneBaseBinding {
  lane: string;
  base: string;
}

export interface EnvironmentPromotionTopology {
  draft_revision: number;
  published_revision: number;
  edges: EnvironmentPromotionEdge[];
  lane_base_bindings: LaneBaseBinding[];
  modify_by?: string;
  modify_time?: string;
}

export interface TopologyValidationIssue {
  code: string;
  path?: string;
  message: string;
}

const normalizeTopology = (raw?: Partial<EnvironmentPromotionTopology>): EnvironmentPromotionTopology => ({
  draft_revision: Number(raw?.draft_revision || 0),
  published_revision: Number(raw?.published_revision || 0),
  edges: raw?.edges || [],
  lane_base_bindings: raw?.lane_base_bindings || [],
  modify_by: raw?.modify_by,
  modify_time: raw?.modify_time,
});

export async function describeEnvironmentPromotionTopology() {
  return normalizeTopology(await getApiRequest<EnvironmentPromotionTopology>({ action: TOPOLOGY_URL }));
}

export async function validateEnvironmentPromotionTopology(topology: EnvironmentPromotionTopology) {
  return apiRequest<{ valid: boolean; issues: TopologyValidationIssue[] }>({
    action: `${TOPOLOGY_URL}/validate`,
    data: topology,
  });
}

export async function saveEnvironmentPromotionTopology(topology: EnvironmentPromotionTopology) {
  return normalizeTopology(await putApiRequest<EnvironmentPromotionTopology>({
    action: TOPOLOGY_URL,
    data: topology,
  }));
}

export async function publishEnvironmentPromotionTopology(expectedRevision: number, comment: string) {
  return apiRequest({
    action: `${TOPOLOGY_URL}/publish`,
    data: { expected_revision: expectedRevision, comment },
  });
}
